### ISSUE TITLE
Nested Queries in 3.8.22 not working

### ISSUE THREAD

**Issue by mmerkes** [LearnBoost/mongoose#2648]
I did a clean npm install in our project, and some tests that were previously working are now broke. I don't know what previous version I was on, but my best guess would be 3.8.21 or earlier.

Previously, I could build a query with the following syntax:

    MyModel.find({
      some_param: 'param',
      some_obj: {
        param_a: 'a',
        param_b: 'b'
      }
    }, callback);

That no longer works in 3.8.22. However, the following syntax does work:

    MyModel.find({
      some_param: 'param',
      'some_ob.param_a': 'a',
      'some_obj.param_b': 'b'
    }, callback);

While it still works, it's a big inconvenience when you've got nested objects. When I have a chance, I will try to create a bare-bones repro.

**Comment by vkarpov15** [comment]
Hmm I don't think this is the exact cause. In MongoDB,

```
MyModel.find({
  some_param: 'param',
  some_obj: {
    param_a: 'a',
    param_b: 'b'
  }
}, callback);
```

Is only supposed to match documents where `some_obj` is exactly equal to `{ param_a: 'a', param_b: 'b' }`, with no other properties. For instance, the following standalone script fails on mongoose 3.6.17, 3.8.1, 3.8.22, 4.0.0-rc1, etc:

```
var mongoose = require('mongoose');
var assert = require('assert');
var Schema = mongoose.Schema;

mongoose.connect('mongodb://localhost:27017/gh-2648');

var schema = Schema({ food: { eggs: Number, bacon: Number } });
var Breakfast = mongoose.model('gh-2648', schema, 'gh-2648');

Breakfast.create({ food: { eggs: 4, bacon: 2 } }, function(error) {
  assert.ifError(error);
  var query = { food: { eggs: 4 } };
  Breakfast.findOne(query, function(error, doc) {
    assert.ifError(error);
    assert.ok(doc);
    assert.equal(4, doc.food.eggs);
    assert.equal(2, doc.food.bacon);
    process.exit(0);
  });
});
```

However, once you make the doc an exact match, like below:

```
var mongoose = require('mongoose');
var assert = require('assert');
var Schema = mongoose.Schema;

mongoose.connect('mongodb://localhost:27017/gh-2648');

var schema = Schema({ food: { eggs: Number, bacon: Number } });
var Breakfast = mongoose.model('gh-2648', schema, 'gh-2648');

Breakfast.create({ food: { eggs: 4, bacon: 2 } }, function(error) {
  assert.ifError(error);
  var query = { food: { eggs: 4, bacon: 2 } };
  Breakfast.findOne(query, function(error, doc) {
    assert.ifError(error);
    assert.ok(doc);
    assert.equal(4, doc.food.eggs);
    assert.equal(2, doc.food.bacon);
    process.exit(0);
  });
});
```

You get the correct behavior on 3.6.x, 3.8.x, and 4.x. So I suspect the reason why your tests are failing is something different. Can you tell me how you're creating the objects that you're querying, and what they look like in the db?

**Comment by mmerkes** [comment]
@vkarpov15 I appreciate pointing out that sub-documents must match exactly, which I didn't realize before. It will assuredly help me figure out future bugs.

I did a bunch of playing around, and I haven't been able to fully and consistently reproduce the issue. As I played around with it more, my issue may be related to having nested refs that were not getting cast to ObjectIds when getting passed a string. My recollection is that I tried passing in strings and strings cast to ObjectIds, which neither worked, though I can't guarantee that I did everything correctly.

Since I can't repro, I'll close the issue. Thanks for your help. If I can repro my issue in the future, I'll come back.
